Add FAQ entries for common syntax questions (#2011)
carbon-language/carbon-lang: docs/project/faq.md

solution for C++ use-cases that require its full performance, low-level control,
and access to hardware.

> ## Why is Carbon designed that way?

Reviewer: ```suggestion
## How were specific feature designs chosen?
```

Small suggestion: "that way" rings a bit disapproving to me, whereas this would be more something we could respond with:
"Throughout the design, we include 'Alternatives considered' and 'References' sections which can be used to research the decision process for a particular design." (like the second paragraph at https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/faq.md#what-can-i-do-if-i-disagree-with-a-design-decision)

Author: Done.

> ### Why are there no angle brackets?

Reviewer: ```suggestion
### Why are parentheses used for templates and generics instead of angle brackets?
```

Angle brackets exist for comparison operators; it's probably going to be helpful to be specific about the feature. But also, I think it can be helpful to be more positive in the phrasing of questions.

Author: > Angle brackets exist for comparison operators

To me, "angle bracket" refers specifically to using those tokens as part of a balanced-pair syntax, but I've rephrased to avoid the confusion.

> it's probably going to be helpful to be specific about the feature.

I'm trying to explain why we don't use angle brackets _anywhere_, not just why we don't use them for templates and generics.

> I think it can be helpful to be more positive in the phrasing of questions.

I don't disagree, but in light of the above I don't see a positive way to phrase this question.

Angle brackets cause major parsing problems in a language that also uses `<` and
`>` as comparison operators. In particular,
[one of our sub-goals for Carbon](/docs/project/goals.md#fast-and-scalable-development)
is that it should support parsing without contextual or semantic information,
and experience with C++ has shown that angle brackets make that difficult if not

> impossible.

Reviewer: ```suggestion
[One of our goals for Carbon](/docs/project/goals.md#fast-and-scalable-development)
is that it should support parsing without contextual or semantic information,
and experience with C++ has shown that using angle brackets for both templates and comparison operators makes that difficult.
```

Focus on objective critiques. Don't exaggerate the difficulty of parsing C++ code by saying it is "impossible"; this is part of maintaining a kind community.

Also, keep an eye on explaining what the use _other_ than comparison operators is: again, that's not explained here.

Author: > Also, keep an eye on explaining what the use _other_ than comparison operators is: again, that's not explained here.

As above, that's intentional; I don't want to focus on templates because I want to explain why we can't use angle brackets anywhere else either.

I think I've addressed the rest of this comment.

For example, in C++, the expression `a<b>(c)` could parse as either a function
call with a template argument `b` and an ordinary argument `c`, or as a chained
comparison `(a < b) > (c)`. In order to resolve the ambiguity, the compiler has
to perform name lookup on `a` to determine whether there's a function named `a`
in scope.

It's also worth noting that Carbon doesn't use _any_ kind of brackets to mark
template parameters (or generic parameters), so if Carbon had angle brackets,
they would mean something different than they do in C++, which could cause
confusion. We do use square brackets to mark _deduced_ parameters, as in:

```

> fn Sort[T:! Comparable](a: Vector(T)*)

Reviewer: Do you think it's worth providing example of template parameter syntax?

Author: I think that would be too much of a digression from the main point here.

```

> But those aren't the same thing. In particular, deduced parameters are never

Reviewer: Same thing as what? I'm not sure what this is trying to say; can it be removed?

Author: How's this?

mentioned at the callsite, so those square brackets are never part of the
expression syntax.

See [this issue](https://github.com/carbon-language/carbon-lang/issues/565) for

> more background on how and why we chose our current generics syntax.

Reviewer: Per some digging, proposal #676 may be a better resource, particularly since we say issues for leads are provisional. It comments on `<>` in the rationale in particular.

Author: Thanks! Done.

### Why aren't variable declarations more concise?

In Carbon, a declaration of a single variable looks like this:

```
var the_answer: i32 = 42;
```

> But this is actually a special case. The syntax between `var` and `=` can be any

Reviewer: ```suggestion
But this is just one case of variable declaration syntax. The syntax between `var` and `=` can be any
```

"actually a special case" implies "actually rare" to me. Instead calling it "the most common case" would also be fair. Or you could change this more significantly, and replace this entire paragraph with "However, the syntax also supports declaring multiple variables together, such as for tuple destructuring:"

Author: I went with "most common case"

[irrefutable pattern](/docs/design/README.md#patterns), not just a single
variable binding. For example:

```
var ((x: i32, _: i32), y: auto) = ((1, 2), (3, 4));
```

This code is valid, and initializes `x` to `1` and `z` to `(3, 4)`. In the
future, we will probably also support destructuring structs in a similar way,
and many other kinds of patterns are possible.

As a result we need `var` to act as an introducer token, so that the parser can
tell it's parsing a variable declaration without unbounded lookahead, which we
think is

> [important to avoid](/docs/project/goals.md#fast-and-scalable-development).

Reviewer: While you start with "As a result", I think the "why" is actually "without unbounded lookahead". I'd suggest walking readers through the example a little more, e.g.:

Now consider the same code without `var`:

```
((x: i32, _: i32), y: auto) = ((1, 2), (3, 4));
```

Whereas with `var` the parser knows a variable will follow, reaching the same conclusion without `var` requires looking through the statement until at least the `=`, and it's easy for typos to confuse that parse. This is unbounded lookahead, which we
think is
[important to avoid](/docs/project/goals.md#fast-and-scalable-development).

Author: How's this?

> `the_answer: i32` is an example of a _binding pattern_, which matches any value

Reviewer: Consider a separate "Why do variable declarations require a type?" question for this. Right now I feel like the topic shifts without introducing why the topic is shifting.

Author: How's this?

of the appropriate type, and binds the given name to it. The `: i32` can't be
omitted, because `the_answer` on its own is an expression, and any Carbon
expression is also a valid pattern, which matches if the value being matched is
equal to the value of the expression. So `var the_answer = 42;` would try to
match `42` with the value of the expression `the_answer`, which requires a

> variable named `the_answer` to already exist.

Reviewer: FWIW, this rationale feels a little unsatisfactory to me. Above, you explain that `var` is used in order to avoid requiring lookahead to figure out what's going on. So we already know this is a variable declaration: why not just say it should never try to match the expression of `the_answer`?

But, noting it's provisional, it may be that there's not a more satisfactory answer right now.

Author: > why not just say it should never try to match the expression of `the_answer`?

The next paragraph is trying to answer that question. Any suggestions for how to make that clearer?

Reviewer: Looking and working to rephrase as a suggestion comment.

solution for C++ use-cases that require its full performance, low-level control,
and access to hardware.

## Why is Carbon designed that way?

### Why aren't `<` and `>` used as delimiters?

[One of our goals for Carbon](/docs/project/goals.md#fast-and-scalable-development)
is that it should support parsing without contextual or semantic information,
and experience with C++ has shown that using `<` as both a binary operator and
an opening delimiter makes that goal difficult to achieve.

For example, in C++, the expression `a<b>(c)` could parse as either a function
call with a template argument `b` and an ordinary argument `c`, or as a chained
comparison `(a < b) > (c)`. In order to resolve the ambiguity, the compiler has
to perform name lookup on `a` to determine whether there's a function named `a`
in scope.

It's also worth noting that Carbon doesn't use _any_ kind of brackets to mark

> template parameters (or generic parameters), so if Carbon had angle brackets,

Reviewer: ```suggestion
[template parameters (and also generic parameters)](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/README.md#checked-and-template-parameters), so if Carbon had angle brackets,
```

To my below comment, which I've marked resolved, I do think there should be some link/reference for what _is_ used for template parameters. But maybe this reflects your intent?

(also, suggesting replacing "or" with "and also" to make clear to newcomers that "template" and "generic" aren't two terms for the same thing)

Author: I've added the link, in a slightly different place.

> (also, suggesting replacing "or" with "and also" to make clear to newcomers that "template" and "generic" aren't two terms for the same thing)

That feels awkward because we're inside a negation, and "not (A and B)" conveys less information than "not (A or B)". How about this instead?

they would mean something different than they do in C++, which could cause
confusion. We do use square brackets to mark _deduced_ parameters, as in:

```
fn Sort[T:! Comparable](a: Vector(T)*)
```

But deduced parameters aren't the same thing as template parameters. In
particular, deduced parameters are never mentioned at the callsite, so those
square brackets are never part of the expression syntax.

> See [p0676](/proposals/p0676.md) for more background on how and why we chose our

Reviewer: ```suggestion
See [Proposal #676: `:!` generic syntax](/proposals/p0676.md) for more background on how and why we chose our
```

```
((_: i32, x: i32), y: auto) = ((1, 2), (3, 4));
```

With this example, the parser would need to look four tokens ahead to determine
that it's parsing a variable declaration rather than an expression. With more
deeply-nested patterns, it would have to look ahead farther. Avoiding this sort
of unbounded lookahead is an important part of our
[fast and scalable development](/docs/project/goals.md#fast-and-scalable-development)
goal.

### Why do variable declarations have to have types?

As discussed above, Carbon variable declarations are actually doing a form of
pattern matching. In a declaration like this:

```
var the_answer: i32 = 42;
```

`the_answer: i32` is an example of a _binding pattern_, which matches any value
of the appropriate type, and binds the given name to it. The `: i32` can't be
omitted, because `the_answer` on its own is an expression, and any Carbon
expression is also a valid pattern, which matches if the value being matched is
equal to the value of the expression. So `var the_answer = 42;` would try to
match `42` with the value of the expression `the_answer`, which requires a
variable named `the_answer` to already exist.

> There are other ways of approaching pattern matching, but they all seem to have
> significant tradeoffs. For example, in Rust, identifiers inside patterns are
> always treated as new name bindings, not expressions that mention existing
> names, but that means the same syntax can have very different meanings depending
> on the context it appears in, which can create confusion. Leaving variable types
> unmarked by default may also have readability drawbacks.
> 
> As of August 2022, the design for pattern matching is provisional, and hasn't
> been through our [evolution process](evolution.md). A future proposal for
> pattern matching will need to weigh these tradeoffs in more detail, and may come
> to a different decision.

Reviewer: ```suggestion
There are other ways of approaching pattern matching, but there are tradeoffs. Pattern matching is still on a provisional design, and as of August 2022 it hasn't been fully reviewed with alternatives considered. A future proposal for pattern matching will need to weigh the tradeoffs in more detail, and may come to a different decision.
```

Reviewer: Suggesting a "less is more" approach: if there's no detailed analysis to link to, leave it at that.
